### PR TITLE
[WIP] Unique dynamic function for time dependent problem

### PR DESCRIPTION
**Pull request by Kev1CO**
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document [docs/contribution.md]?
* [ ] Have you checked to ensure there aren't other open [Pull Requests] for the same update/change?
* [ ] Have you opened/linked the issue related to your pull request?
* [ ] Have you used the tag [WIP] for on-going changes, and removed it when the pull request was ready?
* [ ] When ready to merge, have you sent a comment pinging @pariterre in it?

### New Feature Submissions:

1. [ ] Does your submission pass the tests (if not please explain why this is intended)?
2. [ ] Did you write a proper documentation (docstrings and ReadMe)
3. [ ] Have you linted your code locally prior to submission (using the command: `black . -l120 --exclude "external/*"`)?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new examples for your core changes, as applicable?
* [ ] Have you written new tests for your core changes, as applicable?

- - -
This change is [Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/718)

### CODE REVIEW

**Comment by Ipuch** [comment]
The should be no variable container for time, Proof:

```
f(t,x,u) = t * x + u
p = (tf0, tf1)
ns= 2
phase = 2
phase 0
f(0, x, u)
f(p(0)/2 x, u)
f(p(0))
phase 1
f(p(0), x, u)
f(p(0) + p(1)/2, x, u)
f(p(0) + p(1), x, u)
```

**Review by pariterre — CHANGES_REQUESTED** [review]
GG so far!

> Reviewed 27 of 41 files at r1, 14 of 14 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/718#-:-NbUo7EUEO_saWdrAjet:b2m0f5f)* status: all files reviewed, 11 unresolved discussions (waiting on @Kev1CO)

___
*[`bioptim/dynamics/configure_problem.py` line 130 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUb8_r8h1WHcu1Qn6w:-NbUb8_r8h1WHcu1Qn6x:b-i6pag8) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/dynamics/configure_problem.py#L130)):*
> ```Python
>             A reference to the ocp
>         nlp: NonLinearProgram
>             A reference to the phase
> ```

Do not copy docstrings



_Code quote:_
```Python
A reference to the phase
```
___
*[`bioptim/dynamics/integrator.py` line 291 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUh7Yk0SyDhPTQ82AU:-NbUh7Yk0SyDhPTQ82AV:b2bpbgc) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/dynamics/integrator.py#L291)):*
> ```Python
>             t_norm_init = (i-1)/self.n_step if self.control_type == ControlType.LINEAR_CONTINUOUS else ((self.t_span[1]-self.t_span[0])*(i-1))/self.n_step
>             # x[:, i] = self.next_x(h, t_norm_init, x[:, i - 1], u, p, s)
>             x[:, i] = self.next_x(h, t, x[:, i - 1], u, p, s)
> ```

This should be the vanilla signature of `next_x` from now on (adjust `get_u` accordingly)



_Code quote:_
```Python
x[:, i] = self.next_x(h, t, x[:, i - 1], u, p, s)
```
___
*[`bioptim/dynamics/ode_solver.py` line 129 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUhgLL5n8CBwWUG3hd:-NbUhgLL5n8CBwWUG3he:b-akoc5w) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/dynamics/ode_solver.py#L129)):*
> ```Python
> 
>         ode_opt = {
>             "t0": ocp.node_time(phase_idx=nlp.phase_idx, node_idx=node_index),
> ```

You will probably need to add node_index here
___
*[`bioptim/examples/muscle_driven_ocp/muscle_activations_tracker.py` line 167 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiNlECcgITuQHdZQC:-NbUiNlECcgITuQHdZQD:b-nxfapt) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/examples/muscle_driven_ocp/muscle_activations_tracker.py#L167)):*
> ```Python
>         "ForwardDyn",
>         dyn_func(
>             time=MX(),
> ```

This should be symbolic
___
*[`bioptim/gui/plot.py` line 357 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiDFt1DTuTGNSecP8:-NbUiDFt1DTuTGNSecP9:bxq4key) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/gui/plot.py#L357)):*
> ```Python
>                             if nlp.plot[key].parameters["penalty"].multinode_penalty:
>                                 if casadi_function is not None:
>                                     # size time is len(casadi_function.nominal_in(0))
> ```

Do not leave useless comments
___
*[`bioptim/interfaces/interface_utils.py` line 260 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiZOJ7lNBGomBaQdz:-NbUiZOJ7lNBGomBaQe-:bj5i0qk) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/interfaces/interface_utils.py#L260)):*
> ```Python
>         else:
>             if is_unscaled:
>                 # TODO : add _t like _x for every if unscaled
> ```

to be removed?
___
*[`bioptim/limits/penalty.py` line 1009 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUikSQEOmrixu1RBr9:-NbUikSQEOmrixu1RBrA:b8y9o5m) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/limits/penalty.py#L1009)):*
> ```Python
>                 #     t0=controller.ocp.node_time(phase_idx=controller.get_nlp.phase_idx, node_idx=controller.node_index),
>                 #     x0=controller.states.cx_start, p=u, params=controller.parameters.cx)[
>                 #     "xf"]
> ```

Do not comment when not needed
___
*[`bioptim/limits/penalty_controller.py` line 219 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUj_GNCFMwoCtDEoIh:-NbUj_GNCFMwoCtDEoIi:b1fcf5k) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/limits/penalty_controller.py#L219)):*
> ```Python
>         out = self._nlp.time.scaled
>         out.current_cx_to_get = self.cx_index_to_get
>         return out
> ```

I think this should be removed



_Code quote:_
```Python
    @property
    def time_scaled(self) -> OptimizationVariableList:
        """
        Return the scaled states associated with the current node index.

        Warning: Most of the time, the user does not want that states but the normal `states`, that said, it can
        sometime be useful for very limited number of use case.

        Returns
        -------
        The scaled states at node node_index
        """
        self._nlp.time.node_index = self.node_index
        out = self._nlp.time.scaled
        out.current_cx_to_get = self.cx_index_to_get
        return out
```
___
*[`bioptim/optimization/non_linear_program.py` line 154 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUjl8U7SzAoobUnx_Z:-NbUjl8U7SzAoobUnx__:b4nz137) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/optimization/non_linear_program.py#L154)):*
> ```Python
>         self.plot_mapping = {}
>         self.T = None
>         self.T_scaled = None
> ```

To be removed



_Code quote:_
```Python
self.T_scaled = None
```
___
*[`bioptim/optimization/optimization_vector.py` line 504 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUk9F6AOA5tpTRpbe1:-NbUk9F6AOA5tpTRpbe2:b2xe7y1) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/optimization/optimization_vector.py#L504)):*
> ```Python
>                     s_array = v_array[offset : offset + nstochastic].reshape(
>                         (nlp.stochastic_variables.shape, -1), order="F"
>                     )  # @pariterre "F" seems like an interpolation?
> ```

Nope, this is a numpy nomenclature for `Fotran`



_Code quote:_
```Python
# @pariterre "F" seems like an interpolation?
```
___
*[`bioptim/optimization/solution.py` line 506 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUmq9P5wqpf9vTvlug:-NbUmq9Q3jK11DkCQipM:b-kxunpv) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/optimization/solution.py#L506)):*
> ```Python
> 
>             self.vector = np.ndarray((0, 1))
>             sol_time, sol_states, sol_controls, sol_stochastic_variables = _sol[3], _sol[0], _sol[1], _sol[2]
> ```

Put time at index 3, and move the `For time` block below accordingly (also move params here, see suggestion)



_Suggestion:_
```Python
sol_states, sol_controls, sol_params, sol_time, sol_stochastic_variables = _sol
```

**Comment by Kev1CO** [comment]
*[`bioptim/dynamics/ode_solver.py` line 129 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUhgLL5n8CBwWUG3hd:-Nbnub8V8pgPh5mAGlal:bgyobut) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/dynamics/ode_solver.py#L129)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

You will probably need to add node_index here
</blockquote></details>

Found another way for get_u not requiering node_index.

**Review by Kev1CO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/718)* status: 33 of 42 files reviewed, 11 unresolved discussions (waiting on @pariterre)

___
*[`bioptim/dynamics/configure_problem.py` line 130 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUb8_r8h1WHcu1Qn6w:-Nbnu93SE0__Uq61Sy5O:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/dynamics/configure_problem.py#L130)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Do not copy docstrings
</blockquote></details>

Done.
___
*[`bioptim/dynamics/integrator.py` line 291 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUh7Yk0SyDhPTQ82AU:-NbnuHh06a2JkQv_60ax:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/dynamics/integrator.py#L291)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

This should be the vanilla signature of `next_x` from now on (adjust `get_u` accordingly)
</blockquote></details>

Done.
___
*[`bioptim/examples/muscle_driven_ocp/muscle_activations_tracker.py` line 167 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiNlECcgITuQHdZQC:-NbnvCPaC-FswbOYCRdL:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/examples/muscle_driven_ocp/muscle_activations_tracker.py#L167)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

This should be symbolic
</blockquote></details>

Done.
___
*[`bioptim/gui/plot.py` line 357 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiDFt1DTuTGNSecP8:-NbnvV4x1DhqZwuWqLOl:b-yhg96u) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/gui/plot.py#L357)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Do not leave useless comments
</blockquote></details>

Though it relevent to explain why nominal_in starts from 1 instead of 0.
But yet, removed.
Done.
___
*[`bioptim/interfaces/interface_utils.py` line 260 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiZOJ7lNBGomBaQdz:-Nbnw2yVECUIwgDdEPBC:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/interfaces/interface_utils.py#L260)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

to be removed?
</blockquote></details>

Done.
___
*[`bioptim/limits/penalty.py` line 1009 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUikSQEOmrixu1RBr9:-Nbnwcz1E_RpNi6bw81M:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/limits/penalty.py#L1009)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Do not comment when not needed
</blockquote></details>

Done.
___
*[`bioptim/limits/penalty_controller.py` line 219 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUj_GNCFMwoCtDEoIh:-Nbnwj_Z8a3HN04WnhF2:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/limits/penalty_controller.py#L219)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

I think this should be removed
</blockquote></details>

Done.
___
*[`bioptim/optimization/non_linear_program.py` line 154 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUjl8U7SzAoobUnx_Z:-Nbnx7MM4k69skGQcX8P:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/optimization/non_linear_program.py#L154)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

To be removed
</blockquote></details>

Done.
___
*[`bioptim/optimization/optimization_vector.py` line 504 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUk9F6AOA5tpTRpbe1:-NbnxFTg7aJ1YCtjAinm:b-34mrx) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/optimization/optimization_vector.py#L504)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Nope, this is a numpy nomenclature for `Fotran`
</blockquote></details>

From previous stochastic PR, when merged.
Done.
___
*[`bioptim/optimization/solution.py` line 506 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUmq9P5wqpf9vTvlug:-Nbny0IT7B45DEzOMDKX:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/optimization/solution.py#L506)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Put time at index 3, and move the `For time` block below accordingly (also move params here, see suggestion)
</blockquote></details>

Done.

**Comment by Kev1CO** [comment]
Even though we have discussed it, for me, it is not logical to swap time position from dynamics to solution:
Asked : Solution(solution_ocp, [x_init, u_init_for_solution, p_init, t_init, s_init])
Proposed : Solution(solution_ocp, [t_init, x_init, u_init_for_solution, p_init, s_init])

**Comment by Kev1CO** [comment]
Correct if param is a int in nlp l247

**Comment by Kev1CO** [comment]
Done.
Need to confirm fix comment on pr review

**Comment by Kev1CO** [comment]
Known errors to correct before merge :
- [] Test mhe : control tau is considered as a state key after second iteration
- [] All np.assert tests : The values are the same but array 'out' position is different
- [] Test control_type_none : All np.assert are wrong as the time t in the integration wasn't calculated correctly before this PR
- [] Test time_dependent_problems : np.assert arn't done yet as the PR must be correct before completing values

File: bioptim/examples/stochastic_optimal_control/arm_reaching_muscle_driven.py [hunk 1]
```diff
@@ -608,9 +617,11 @@ def prepare_socp(
         dynamics,
         n_shooting,
         final_time,
+        t_init=t_init,
```
Comment (Kev1CO) [inline bioptim/examples/stochastic_optimal_control/arm_reaching_muscle_driven.py:620]: Relevence of having initial guess and bound for time as it is driven by ocp.node_time ?
Was implemented when time was variable container

File: bioptim/examples/stochastic_optimal_control/arm_reaching_torque_driven_explicit.py
Comment (Kev1CO) [inline bioptim/examples/stochastic_optimal_control/arm_reaching_torque_driven_explicit.py:1]: Not tested is it normal ?

File: bioptim/examples/torque_driven_ocp/example_pendulum_time_dependent.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    # --- Print ocp structure --- #
+    ocp.print(to_console=False, to_graph=False)
+
+    # --- Solve the ocp --- #
+    sol = ocp.solve(Solver.IPOPT(show_online_optim=platform.system() == "Linux"))
+    sol.graphs(show_bounds=True)
+
+    # --- Show the results in a bioviz animation --- #
+    sol.print_cost()
+    # sol.animate(n_frames=100)  # Question : sol.animate crashes when assume phase dynamic is False
```
Comment (Kev1CO) [inline bioptim/examples/torque_driven_ocp/example_pendulum_time_dependent.py]: Sol.animate crashes when assume phase dynamics = false

File: bioptim/optimization/optimal_control_program.py [hunk 3]
```diff
@@ -140,16 +143,19 @@ class OptimalControlProgram:
     """
 
     # TODO: OCP should not be aware of s (s_init, s_bounds...)
+    #  OCP should not be aware of t (t_init, t_bounds...) ???
```
Comment (Kev1CO) [inline bioptim/optimization/optimal_control_program.py:143]: Bounds and initial in ocp ?

File: tests/shard1/test_controltype_none.py [hunk 4]
```diff
@@ -258,6 +230,7 @@ def test_main_control_type_none(use_sx, assume_phase_dynamics):
     """
 
     # TODO It seems assume_phase_dynamics=True is broken
```
Comment (Kev1CO) [inline tests/shard1/test_controltype_none.py]: Not broken, dynamics changes at each node, true isn't suppose to work

File: tests/shard1/test_dynamics.py [hunk 5]
```diff
@@ -90,14 +90,15 @@ def test_torque_driven(with_contact, with_external_force, cx, rigidbody_dynamics
     ConfigureProblem.initialize(ocp, nlp)
 
     # Test the results
+    time = np.random.rand(1, nlp.ns)
```
Comment (Kev1CO) [inline tests/shard1/test_dynamics.py:93]: not useful now but might be tested in the future if time is included into dynamics
aswell to see if time impacts the result when it shouldn't

File: tests/test_time_dependent_problems.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        assume_phase_dynamics=assume_phase_dynamics,
+    )
+
+
+@pytest.mark.parametrize("n_phase", [1, 2])
+@pytest.mark.parametrize(
+    "integrator",
+    [
+        OdeSolver.RK4(),
+        # OdeSolver.IRK(),  # stuck into infinite loop
```
Comment (Kev1CO) [inline tests/test_time_dependent_problems.py:209]: disabled because stuck into infinit loop

**Review by pariterre — CHANGES_REQUESTED** [review]
> Reviewed 22 of 46 files at r5, 14 of 26 files at r6, 16 of 16 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/718)* status: all files reviewed, 16 unresolved discussions (waiting on @Kev1CO)

___
*[`bioptim/dynamics/dynamics_functions.py` line 924 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304109169:-NcbjszJ6Fmj3lGFjTXj:b943ayk) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/dynamics/dynamics_functions.py#L924)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Why all MX.sym but holonomic
</blockquote></details>

MX.sym indeed
___
*[`bioptim/dynamics/integrator.py` line 16 at r7](https://reviewable.io/reviews/pyomeca/bioptim/718#-NcbkHNX6zXfyt69TFYH:-NcbkHNX6zXfyt69TFYI:brbzdi6) ([raw file](https://github.com/pyomeca/bioptim/blob/00fb8a58e6e2d007ff74d37f387c370f8511d6ea/bioptim/dynamics/integrator.py#L16)):*
> ```Python
>     model: BioModel
>         The biorbd model to integrate
>     t_span = tuple[float, ...]
> ```

time_integration_grid
___
*[`bioptim/dynamics/ode_solver.py` line 40 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304132118:-NcbkkMb3ALQ2BJdp9vj:b-m0zar3) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/dynamics/ode_solver.py#L40)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Relevence of having t into integration def as it is replaced by t_span during integration
</blockquote></details>

If it is not useful, then remove
___
*[`bioptim/examples/stochastic_optimal_control/arm_reaching_muscle_driven.py` line 620 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304169284:-Ncbm0CY2jSBR1FuhHWV:b-95abdx) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/examples/stochastic_optimal_control/arm_reaching_muscle_driven.py#L620)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Relevence of having initial guess and bound for time as it is driven by ocp.node_time ?
Was implemented when time was variable container
</blockquote></details>

No relevence
___
*[`bioptim/examples/torque_driven_ocp/example_pendulum_time_dependent.py` line 210 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304184499:-NcbmepnCNhTHJa3UkIN:bu6wwx4) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/examples/torque_driven_ocp/example_pendulum_time_dependent.py#L210)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Sol.animate crashes when assume phase dynamics = false
</blockquote></details>

Please add a "TODO Pariterre -> PROBLEM EXPLANATIOn" in the animate method
___
*[`bioptim/gui/plot.py` line 357 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiDFt1DTuTGNSecP8:-Ncbn6d84_TYjHKtw4W7:b6y7eqo) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/gui/plot.py#L357)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Though it relevent to explain why nominal_in starts from 1 instead of 0.
But yet, removed.
Done.
</blockquote></details>

It should indeed stays there
___
*[`bioptim/optimization/non_linear_program.py` line 257 at r7](https://reviewable.io/reviews/pyomeca/bioptim/718#-Ncbp6Vi3nQHkOrsaGXl:-Ncbp6Vi3nQHkOrsaGXm:b-tbjxuj) ([raw file](https://github.com/pyomeca/bioptim/blob/00fb8a58e6e2d007ff74d37f387c370f8511d6ea/bioptim/optimization/non_linear_program.py#L257)):*
> ```Python
>                         f"to the number of phases({ocp.n_phases})."
>                         f"List length of model, final time and node shooting must be equivalent to phase number"
>                     )
> ```

f"...{1 if isinstance(param, int) else len(param)} ..."



_Code quote:_
```Python
                    raise RuntimeError(
                        f"{param_name} size({len(param)}) does not correspond "
                        f"to the number of phases({ocp.n_phases})."
                        f"List length of model, final time and node shooting must be equivalent to phase number"
                    )
```
___
*[`bioptim/optimization/optimal_control_program.py` line 146 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304198790:-NcbpOpVAeFNCWec2Ovz:b-9qdp7i) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/optimization/optimal_control_program.py#L146)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Bounds and initial in ocp ?
</blockquote></details>

not here
___
*[`bioptim/optimization/solution.py` line 502 at r7](https://reviewable.io/reviews/pyomeca/bioptim/718#-NcbpddR9evYZ8tAsrRf:-NcbpddR9evYZ8tAsrRg:bnowi6x) ([raw file](https://github.com/pyomeca/bioptim/blob/00fb8a58e6e2d007ff74d37f387c370f8511d6ea/bioptim/optimization/solution.py#L502)):*
> ```Python
>             self.vector = np.ndarray((0, 1))
>             sol_time, sol_states, sol_controls, sol_params, sol_stochastic_variables = (
>                 _sol[0],
> ```

= _sol
___
*[`tests/test_time_dependent_problems.py` line 209 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304212403:-Ncbpqk149ORSG3YYzDl:b-ucvvy8) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/tests/test_time_dependent_problems.py#L209)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

disabled because stuck into infinit loop 
</blockquote></details>

Enable, but raise at first line if isinstance(integrator, OdeSolver.IRK())
___
*[`tests/shard1/test_dynamics.py` line 93 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304208419:-NcbqpDv4H5zN_ViGCxD:b34qt7s) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/tests/shard1/test_dynamics.py#L93)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

not useful now but might be tested in the future if time is included into dynamics
aswell to see if time impacts the result when it shouldn't
</blockquote></details>

Explain why it is relevent

**Comment by Kev1CO** [comment]
*[`tests/test_time_dependent_problems.py` line 209 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304212403:-NcbvOPS7qc6610SkWos:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/tests/test_time_dependent_problems.py#L209)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Enable, but raise at first line if isinstance(integrator, OdeSolver.IRK())
</blockquote></details>

Done.

**Comment by Kev1CO** [comment]
*[`bioptim/dynamics/ode_solver.py` line 40 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304132118:-Ncbw4_0E5KLWTJpfLZ5:b-7rl0jd) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/dynamics/ode_solver.py#L40)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

If it is not useful, then remove
</blockquote></details>

done.

**Comment by Kev1CO** [comment]
*[`bioptim/examples/stochastic_optimal_control/arm_reaching_muscle_driven.py` line 620 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304169284:-NcbwQrZ6fHuGwCdqPzx:b-7rl0jd) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/examples/stochastic_optimal_control/arm_reaching_muscle_driven.py#L620)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

No relevence
</blockquote></details>

done.

**Comment by Kev1CO** [comment]
*[`bioptim/examples/torque_driven_ocp/example_pendulum_time_dependent.py` line 210 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304184499:-NcbxI5gD5KMZyL0WnM0:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/examples/torque_driven_ocp/example_pendulum_time_dependent.py#L210)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Please add a "TODO Pariterre -> PROBLEM EXPLANATIOn" in the animate method
</blockquote></details>

Done.

**Comment by Kev1CO** [comment]
*[`tests/shard1/test_dynamics.py` line 93 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304208419:-Ncbyz1t0kUkwz-DoPIh:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/tests/shard1/test_dynamics.py#L93)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Explain why it is relevent
</blockquote></details>

Done.

**Review by Kev1CO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/718)* status: all files reviewed, 7 unresolved discussions (waiting on @pariterre)

___
*[`bioptim/dynamics/integrator.py` line 16 at r7](https://reviewable.io/reviews/pyomeca/bioptim/718#-NcbkHNX6zXfyt69TFYH:-NcbrzlD4ShPpSG3YQQu:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/00fb8a58e6e2d007ff74d37f387c370f8511d6ea/bioptim/dynamics/integrator.py#L16)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

time_integration_grid
</blockquote></details>

Done.
___
*[`bioptim/gui/plot.py` line 357 at r2](https://reviewable.io/reviews/pyomeca/bioptim/718#-NbUiDFt1DTuTGNSecP8:-NcbxM8V-3-23F8xiax6:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/b5c59770e1813aab0726e8dab99c42e23c4e2ec7/bioptim/gui/plot.py#L357)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

It should indeed stays there
</blockquote></details>

Done.
___
*[`bioptim/optimization/non_linear_program.py` line 257 at r7](https://reviewable.io/reviews/pyomeca/bioptim/718#-Ncbp6Vi3nQHkOrsaGXl:-Ncbxnpo3QLislVQ9Nol:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/00fb8a58e6e2d007ff74d37f387c370f8511d6ea/bioptim/optimization/non_linear_program.py#L257)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

f"...{1 if isinstance(param, int) else len(param)} ..."
</blockquote></details>

Done.
___
*[`bioptim/optimization/solution.py` line 502 at r7](https://reviewable.io/reviews/pyomeca/bioptim/718#-NcbpddR9evYZ8tAsrRf:-Ncby35165TZohXd6_Ub:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/00fb8a58e6e2d007ff74d37f387c370f8511d6ea/bioptim/optimization/solution.py#L502)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

= _sol
</blockquote></details>

Done.

**Comment by Kev1CO** [comment]
*[`bioptim/dynamics/configure_problem.py` line 3 at r6](https://reviewable.io/reviews/pyomeca/bioptim/718#gh-1304110133:-Ncbz70m4eDyg5P-2TpA:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/223e3a6a5381c479e5f933496904aa35bc88e1cd/bioptim/dynamics/configure_problem.py#L3)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Lot of unused packages
</blockquote></details>

Done.

**Pull request merged**